### PR TITLE
[WIP] FTO behavior

### PR DESCRIPTION
**Pull request by ryzokuken**
Sample Output (As of now):

```
➜  plotsbot git:(minor-2) ✗ TEST=true node src/bot
Bot is running in testing mode.
[ryzokuken => plotsbot-ryzokuken]
fto
1451 => Adjust subscription_mailer to use rusers table
1418 => Add more functional tests to each method in Search API
1405 => Fix "like" list URL on notes and wiki pages
171 => rss feeds empty for banned users
```

@jywarren @ananyo2012 I personally feel that promises are a very crucial part of the bot, because if we do not use promises, the calls would be blocking, which would not allow the bot to work at its full potential.

However, the inclusion of a promise-based workflow makes our job difficult at the same time. Because the promises aren't handled by the behavior itself but the rest of the logic which is completely independent of the bot, it would not understand how to deal with promise resolution and rejection.  
**For eg:** The code from this issue can return a promise which would yield an object. However, the `addMessageHandler` function (the function which actually gets the value) is completely behavior agnostic and has no idea how to use that object to make a valid message.

I welcome your suggestions on this. Here is a possible solution I came up with:

If a particular behavior returns a promise, it is mandatory to also include two other functions inside the object. These two functions, (can be called `onPromiseResolve` and `onPromiseReject`) take the object yielded by the Promise as an argument and return the string that can be sent as message, and takes the error object and takes care of it (log it in our case) respectively. For functions that return a normal string value, it would not be required to define such a function as a default value would be set of these functions which would be good enough only for functions that return a string value and not an object. It could be somewhere on the lines of:

```js
function onPromiseResolve(value) {
  return value;
}
```

**OR BETTER**

```js
const onPromiseResolve = data => data;
```

What do you think about this?

Also, I think it would be more appropriate for behaviors to return classes that extend `Behavior` rather than objects. This way, we would be able to supply some initial input from `bot.js` to the specific behavior through its constructor.  
**Eg:** This particular issue defines an instance of class Github inside the behavior. However, it would be more appropriate to define such an object inside `bot.js`, configure it with the proper credentials once and then pass it as an argument to *ALL* github-based issues.

### CODE REVIEW

**Comment by ananyo2012** [comment]
+1 to using Promises as it uses external api. Yes inheriting different subclasses from the Behavior class can be done. But we have to ensure it will be consistent in long run so that we don't  have to too many API redesigns.

**Comment by jywarren** [comment]
+1!

**Comment by ryzokuken** [comment]
@ananyo2012 I don't think there will be too many API redesigns as I believe that in every redesign, we move forward to make the bot better and more efficient, and there will come a time when it would be at the peak of its productivity. Regarding the other things, I would need your confirmation to proceed. Should I use the model I talked about above? Or should I think about some alternative way?

**Comment by ananyo2012** [comment]
Yes go ahead with the implementation

**Comment by ryzokuken** [comment]
@ananyo2012 @jywarren: Umm, I am both happy and sad to say, that it was a false alarm (for the record I did spend a few hours last night brainstorming about how to handle promise resolutions and rejections). @jkantr highlighted the fact that we could put the logic that would convert the object to a legible message string right inside the behavior, using another set of `then` and `catch` calls (All this happened because I forgot to visualize promise resolution as a chain and not a stack).

Also, apparently we cannot return a descriptive message from the `catch` block to inform the user of what went wrong, so I discussed with @jkantr the possibility of throwing an error inside the catch block inside the behavior and then `catch`ing it again inside the `addMessageHandler` method, thus notifying the user. However, while these "behavior errors" should be passed on to the operator of the bot, all other errors (which will be programmer errors for the most part) should result in a crash. This could be done through a lot of hacky code inside the outer `catch` block, but I think that its high time for us to make use of a more powerful promise implementation, eg: bluebird.

Bluebird will allow us to convert the already hacky looking `Promise.resolve().then(...)` calls into more natural looking `Promise.try(...)` calls and allow for more advanced error handling, somewhere on the lines of:

```js
  addMessageHandler() {
    this.client.addMessageHandler((from, to, message) => {
      this.getResponse(to, message).then((response) => {
        if(response) {
          if(to === this.botNick) {
            // Message was recieved in a DM
            this.client.sendMessage(from, response);
          } else {
            // Message was recieved in a normal channel
            this.client.sendMessage(to, response);
          }
        }
      }).catch(BehaviorError, (err) => {
          // this error was thrown by a behavior action, so the
          //error message is already formulated to be displayed
          this.client.sendMessage(null, err.message);
      }).catch((err) => {
          // these are errors that propegated up the chain from
          // elsewhere uncaught - don't send these back to chat
          console.error(err);
      });
    });
  }
```
@ananyo2012 what do you think? I believe using bluebird will make our lives easy.

**Comment by ryzokuken** [comment]
Let's discuss how to deal with promise rejections in the best manner possible. Meanwhile, I will write code to ensure proper results on promise resolution.

**Comment by ryzokuken** [comment]
@ananyo2012 @jywarren let's decide what to do next?

**Comment by jywarren** [comment]
Sorry, i'm not that familiar with promises, to be honest. Glad you're thinking this through carefully. would `  addMessageHandler() {` be in each behavior, or is this the overall handler for all behaviors? It's just a bit dense to read if it's something in each behavior.

**Comment by ryzokuken** [comment]
@jywarren `addMessageHandler()` is the message inside the `Behaviors` class which adds all the appropriate message handlers to the interface. Maybe I'm misunderstanding you?  
I also left a few messages on #publiclab-testing, it would be awesome if you could just take a look.

Thanks

**Comment by ryzokuken** [comment]
@jywarren I was thinking of writing the tests for this behavior now. How should I go about that? I cannot ensure what the issues at a certain point are. Perhaps I could use the github module again inside the tests to fetch the tests? Or maybe write a mock github object that returns a certain string (hardcoded) and use that in place of the original object?

**Comment by jywarren** [comment]
"write a mock github object that returns a certain string (hardcoded) and
use that in place of the original object?" sounds great.

**Comment by ryzokuken** [comment]
@jywarren Great! Will try to accomplish that soon.

**Comment by ryzokuken** [comment]
@jywarren what we're your thoughts on behaviors being classes instead of objects again? I'm sorry if you told me and I missed them, just wanted to confirm,

**Comment by jywarren** [comment]
I'm honestly not quite sure -- can you summarize the pros and cons? Ananyo,
what do you think?

**Comment by ryzokuken** [comment]
**TL:DR;** version

Pro: We can make a single Github object inside the `bot.js` file and pass it to each Behavior that depends on such an object as a constructor argument. Otherwise, we will have to define a new one for each behavior.

Con: Idk, really. Maybe the fact that we'll export classes instead of objects and use `new HelpBehavior( ... )` instead of just `helpBehavior`?

**Comment by jywarren** [comment]
Sounds like Pro wins!

**Comment by ryzokuken** [comment]
> zsoc: Your program requires exactly 0 classes.  
zsoc: But here we are, lol.

@jywarren I had been having quite a few problems trying to figure out how to make the model work the way we wanted it to. Because there seemed to be no way out, I took the time to have a long discussion with @jkantr and together, we realized that not only was everything more difficult to with classes, they might not be our best option. Instead, we could use an alternate model we discussed in which we maintain a state object that contains each of the crucial objects we need (eg: github, db, plots2, etc) and the behaviors that have any dependencies export a function that takes the state and returns an object instead of a normal object, which means that all our current behaviors will also not be required to be changed.

The API will be changed this way:

```js
// If the behavior requires the state (eg: FTO behavior)
const ftoBehavior = require('./behaviors/fto')(state);

// If the behavior does not require the state (eg: Help behavior)
const helpBehavior = require('./behavior/help');
```

more explanation on this inside @jkantr's example gist: https://gist.github.com/jkantr/1c4a00bc4d18e275614c373b1de1d099#file-js-L20

What are your thoughts about this? I will try implementing this right now, or in the worst case by tomorrow.

**Comment by jywarren** [comment]
Hmm, I'm hesitant as I'm not sure that's a familiar pattern to many new
coders. What would it mean for those seeking to add new interfaces?

**Comment by jywarren** [comment]
I guess how would you explain this to a newcomer is my question. Thanks!

**Comment by ryzokuken** [comment]
@jywarren That would change nothing for interfaces. For behaviors, this would only be necessary for behaviors that require some dependecy. For normal behaviors, all of the old rules apply. Even for the newer behaviors, all you had to do was to wrap the code that needed an object inside a function that takes an object and uses destructuring to obtain the dependencies they need. Then. they can use the dependencies inside the function and use it to return a more appropriate object.

**Comment by ryzokuken** [comment]
I will explain all of this in the README. IMO, this hardly changes anything if all you wanted to do was to make a "normal" behavior. This only comes into play when you want to make a behavior that needs to share its state with other behaviors. Eg: I wanted all GitHub behaviors to share a common `github` object so that we do not need to authenticate inside every behavior again.

**Comment by jkantr** [comment]
So I guess i'll pipe up here since I keep getting @ mentioned o.O

Parameterized modules are a pattern to solve the "stateful dependency" problem that many run into with Express applications (trying to share stateful deps between multiple detatched routers). Here's another example snippet from someone else: [parametric modules](https://gist.github.com/joepie91/e0e5fd2fed8f603b5ae4c8c7f2409138)

The more 'familiar' pattern for this is usually just to export your dep after configuration from some common shared require. The biggest downside to this (relying on require cache for state) is things can get a little hairy, especially when it comes to cyclical dependencies and mutation. I can't tell you the number of times I've seen somewhere swear their thing was 'ignoring configuration options' when in fact they were just requiring the same object into multiple modules and configuring it in each one (the last one would mutate the one previous to it, on load)

tldr; like @ryzokuken said, this doesn't actually change the api of the library itself, it's just a pattern to share/re-use the same stateful, configured things between the entry point of the application and the various behaviors.. think of it as a form of dependency injection

**Comment by ryzokuken** [comment]
Thanks, @jkantr.

File: src/interfaces/cli.js [hunk 1]
```diff
@@ -15,8 +15,7 @@ class CliClient {
   }
 
   sendMessage(channel, message) {
-    console.log(`[${this.nick} => ${channel}]\n${message}`);
-    console.log(`[${channel} => ${this.nick}]`);
+    console.log(message);
```
Comment (ananyo2012) [inline src/interfaces/cli.js]: I guess there is no reason to change this format. It looks prettier this way

Comment (ryzokuken) [inline src/interfaces/cli.js]: @ananyo2012 When we use promises, there are problems with this format because this does not gets printed instantly.

Comment (jkantr) [inline src/interfaces/cli.js]: What would be the problem here? Do you mean the reference to `this.nick` may have mutated? Because channel and message certainly will not have changed...

**Comment by ryzokuken** [comment]
No, I mean this method will not be called instantly like it should have in
the case of a static behavior. That might give the user an impression of a
blocking call. Therefore, I decided to drop this altogether.

File: src/interfaces/cli.js (hunk 1 above)
Comment (ryzokuken) [inline src/interfaces/cli.js]: No, I mean this method will not be called instantly like it should have in
the case of a static behavior. That might give the user an impression of a
blocking call. Therefore, I decided to drop this altogether.

**Comment by ryzokuken** [comment]
@ananyo2012 Working tests added for the FTO Behavior. This looks almost finished from my side. If you and @jywarren could give your thoughts on the suggested architecture, I would add the required documentation and this would be ready for merge.

This PR already got too big for my liking, but I guess that's okay because this allowed us to think about our new "shared resources" model wherein all behaviors in need of dependencies may share certain state with the others.

**Comment by jywarren** [comment]
This looks great and I appreciate the thought and extra feedback, everyone! +1

**Comment by ryzokuken** [comment]
@jywarren Awesome. I will push the documentation today and hopefully merge this with @ananyo2012's consent. Then I'm thinking of deploying this to Tycho. Is it the right time? Or should I wait and add a few more features before finally deploying.

**Comment by jywarren** [comment]
So, this would be another issue, but I'm wondering if there's a way to do a
short term memory to remember if the bot has already said hello to people.
Seeing it operating in the testing channel, I think there might be an easy
way to have it know if someone has been in the room at least since the bot
has been running. This is no good for long term storage but could smooth
interaction a bit for things like the repeated welcoming greeting.

Then we could deploy?

**Comment by ryzokuken** [comment]
@jywarren I will make another PR for it right now. For now, let's use a simple in-memory array over a database? We could use a more solid solution later, but the one you proposed seems simple and still convenient.

File: src/utils.js [hunk 2]
```diff
@@ -6,7 +6,32 @@ function remove(array, element) {
   array.splice(array.indexOf(element), 1);
 }
 
+const mockGithub = {
```
Comment (ananyo2012) [inline src/utils.js]: But this mock in the spec file instead of the utils file as it is just used in tests.

Comment (ryzokuken) [inline src/utils.js]: this would be shared by all the github behaviors' tests.

Comment (ananyo2012) [inline src/utils.js]: So you can make a separate `github-mock-spec.js` file in the spec directory instead of the utils file

Comment (ryzokuken) [inline src/utils.js]: `*-spec.js` are test files. What about making a `test-utils.js` file inside the `spec` folder instead?

**Comment by ryzokuken** [comment]
@jywarren added the documentation.

**Comment by ryzokuken** [comment]
@jywarren This feature has been really fun to make and I think not only adds real value to the bot, but also makes us more flexible when it comes to making future behaviors. Will wait for further review by @ananyo2012 until tomorrow morning, after which I will merge this and deploy it to Tycho.

Thanks.

**Comment by jywarren** [comment]
:-) great to hear!!!

File: src/utils.js (hunk 2 above)
Comment (ananyo2012) [inline src/utils.js]: Yes that can be done. Just do this and it can be merged!

Comment (ryzokuken) [inline src/utils.js]: @ananyo2012 sounds awesome, just a minute.

**Comment by ryzokuken** [comment]
@ananyo2012 here we go :)

**Comment by ryzokuken** [comment]
@ananyo2012 thanks :)

@ananyo2012 @jywarren for the time-being, until we figure out a way to do that automatically, I will run `npm version` manually and push my changes to master. Hope that's not a problem (I will be doing it right now, if there's an issue we could revert later)

**Comment by jywarren** [comment]
I do this manually too on my projects!

Great work folks!
